src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs

```diff
@@ -698,12 +698,14 @@ private async Task<bool> PushTargets(IList<TargetSpecification> targets, TargetE
                     if (buildReason == TargetBuiltReason.BeforeTargets || buildReason == TargetBuiltReason.DependsOn || buildReason == TargetBuiltReason.None)
                     {
                         TargetEntry currentParent = parentTargetEntry;
+                        List<string> parentChain = new List<string>();
                         while (currentParent != null)
                         {
+                            parentChain.Add(currentParent.Name);
                             if (String.Equals(currentParent.Name, targetSpecification.TargetName, StringComparison.OrdinalIgnoreCase))
                             {
                                 // We are already building this target on this request. That's a circular dependency.
-                                ProjectErrorUtilities.ThrowInvalidProject(targetLocation, "CircularDependency", targetSpecification.TargetName);
+                                ProjectErrorUtilities.ThrowInvalidProject(targetLocation, "CircularDependencyInTargetGraphWithVerbosity", targetSpecification.TargetName, parentTargetEntry.Name, buildReason, string.Join("<-", parentChain));
```

Comment:
Did you account for verbosity in any way?

Reply:
improve the format of circular target dependence error message.